packages/protocol/src/certificate/CertificateAuthority.ts

```diff
@@ -187,14 +275,126 @@ export class CertificateAuthority {
+            (Bytes.isBytes(certValues.icacKeyPair) || typeof certValues.icacKeyPair === "object") &&
+            Bytes.isBytes(certValues.icacKeyIdentifier) &&
+            Bytes.isBytes(certValues.icacCertBytes)
+        );
+    }
+
+    #loadFromStorage(certValues: Record<string, unknown>): void {
+        this.#rootCertId = BigInt(certValues.rootCertId as bigint | number);
+        this.#rootKeyPair = PrivateKey(certValues.rootKeyPair as BinaryKeyPair);
+        this.#rootKeyIdentifier = certValues.rootKeyIdentifier as Bytes;
+        this.#rootCertBytes = certValues.rootCertBytes as Bytes;
+        this.#nextCertificateId = BigInt(certValues.nextCertificateId as bigint | number);
+
+        if (this.#isValidStoredIcacCertificate(certValues)) {
+            this.#icacProps = {
+                certId: BigInt(certValues.icacCertId as bigint | number),
+                keyPair: PrivateKey(certValues.icacKeyPair as BinaryKeyPair),
+                keyIdentifier: certValues.icacKeyIdentifier as Bytes,
+                certBytes: certValues.icacCertBytes as Bytes,
+            };
+        }
+    }
+
+    #buildStorageData(): CertificateAuthority.StorageData {
+        const data: CertificateAuthority.StorageData = {
+            rootCertId: this.#rootCertId,
+            rootKeyPair: this.#initializedRootKeyPair.keyPair,
+            rootKeyIdentifier: this.#initializedRootKeyIdentifier,
+            rootCertBytes: this.#initializedRootCertBytes,
+            nextCertificateId: this.#nextCertificateId,
+            intermediateCert: this.#intermediateCert,
+        };
+
+        if (this.#icacProps) {
+            data.icacCertId = this.#icacProps.certId;
+            data.icacKeyPair = this.#icacProps.keyPair.keyPair;
+            data.icacKeyIdentifier = this.#icacProps.keyIdentifier;
+            data.icacCertBytes = this.#icacProps.certBytes;
+        }
+
+        return data;
+    }
+
+    #getSigningParameters(): {
+        issuer: { rcacId: bigint } | { icacId: bigint };
+        signingKey: PrivateKey;
+        authorityKeyId: Bytes;
+    } {
+        if (this.#icacProps) {
+            return {
+                issuer: { icacId: this.#icacProps.certId },
+                signingKey: this.#icacProps.keyPair,
+                authorityKeyId: this.#icacProps.keyIdentifier,
+            };
+        }
+
+        return {
+            issuer: { rcacId: this.#rootCertId },
+            signingKey: this.#initializedRootKeyPair,
+            authorityKeyId: this.#initializedRootKeyIdentifier,
+        };
+    }
+
+    get #initializedRootCertBytes() {
+        if (this.#rootCertBytes === undefined) {
+            throw new InternalError("Root certificate is not installed");
+        }
+        return this.#rootCertBytes;
+    }
+}
+
+interface IcacProps {
+    certId: bigint;
+    keyPair: PrivateKey;
+    keyIdentifier: Bytes;
+    certBytes: Bytes;
 }
 
 export namespace CertificateAuthority {
+    export type PartialConfiguration = {
```

Comment:
Maybe rename to "CreateConfiguration" or "ConfigurationForCreate" because this is only relevant when we create it